http-netty/src/main/java/io/micronaut/http/netty/EventLoopSerializer.java

```diff
@@ -0,0 +1,152 @@
+
+import io.micronaut.core.annotation.Internal;
+import io.netty.util.concurrent.EventExecutor;
+import io.netty.util.concurrent.OrderedEventExecutor;
+
+/**
+ * This class forwards reactive operations onto an {@link io.netty.channel.EventLoop} if they are
+ * called from outside that loop. It avoids unnecessary {@link EventExecutor#execute} calls when
+ * the reactive methods are called inside the loop. All the while it ensures the operations remain
+ * serialized (in the same order).
+ * <p>
+ * Should be used like this:
+ * <pre>
+ *     public void onNext(Object item) {
+ *         if (serializer.executeNow(() -> onNext0(item))) {
+ *             onNext0(item);
+ *         }
+ *     }
+ *
+ *     private void onNext0(Object item) {
+ *         ...
+ *     }
+ * </pre>
+ * <p>
+ * This class is <b>not</b> thread-safe: The invariants for calls to {@link #executeNow} are very
+ * strict. In particular:
+ * <ul>
+ *     <li>There must be no concurrent calls to {@link #executeNow}.</li>
+ *     <li>When {@link #executeNow} returns {@code true}, the subsequent execution of the child
+ *     method ({@code onNext0} in the above example) must fully complete before the next
+ *     {@link #executeNow} call. This ensures that there are no concurrent calls to the child
+ *     method.</li>
+ * </ul>
+ * Both of these invariants are guaranteed by the reactive spec, but may not apply to other use
+ * cases.
+ *
+ * @since 4.4.0
+ * @author Jonas Konrad
+ */
+@Internal
+public final class EventLoopSerializer {
+    /**
+     * This adds some extra checks to find bugs.
+     */
+    private static final boolean STRICT_CHECKING = false;
+
+    private final OrderedEventExecutor loop;
+    /**
+     * Generation assigned to the next task.
+     */
+    private int submitGeneration = 0;
+    /**
+     * Generation of the next task that can be executed immediately. All tasks with a lower
+     * generation count have been fully executed already, with one exception: If the last task
+     * was submitted on the event loop, {@link #executeNow} returned true and the caller may not
+     * have fully executed it yet.
+     */
+    private volatile int runGeneration = 0;
+
+    public EventLoopSerializer(OrderedEventExecutor loop) {
+        this.loop = loop;
+    }
+
+    /**
+     * Determine whether the next step can be executed immediately. Iff this method returns
+     * {@code true}, {@code delayTask} will be ignored and the caller should call the target method
+     * manually. Iff this method returns {@code false}, the caller should take no further action as
+     * {@code delayTask} will be run in the future.
+     *
+     * @param delayTask The task to run if it can't be run immediately
+     * @return {@code true} if the caller should instead run the task immediately
+     */
+    public boolean executeNow(Runnable delayTask) {
+        // pick a generation ID for this task.
+        int generation = submitGeneration++;
+        if (loop.inEventLoop()) {
```

Comment:
Would it be better to implement this with a lock? Immediate execution (or execution in general) is acquiring the lock, if lock is taken the task should be scheduled. IMHO it would be much easier to understand.